text/0129-video-replay-envelope.md

```diff
@@ -0,0 +1,49 @@
+- Start Date: 2024-02-06
+- RFC Type: feature
+- RFC PR: [#129](https://github.com/getsentry/rfcs/pull/129)
+- RFC Status: draft
+
+# Summary
+
+In order to capture video replays and present them to the user, we need to define/decide:
+
+- how to transport the video data to the server
+- how to integrate the video data into the RRWeb JSON format
+- how to combine multiple replay chunks to a single replay session
+
+All of these influence one another and need to be considered together in a single RFC.
+
+Note: SDK-side implementation that is currently being worked on relies on taking screenshots and encoding them to a video.
+This is based on an evaluation where a video has much smaller size than a sequence of images (**TODO fix these: factor of X for 720p video**).
+
+# Motivation
+
+We need this to to capture replays on platforms where it's not possible/feasible to produce an HTML DOM (i.e. the native format supported by RRWeb). For example: mobile apps.
+
+<!-- # Supporting Data -->
+<!-- Metrics to help support your decision (if applicable). -->
+
+# Options Considered
+
+## Using a video, with EnvelopeItem:Attachment
+
+- From the SDK, we should be able to use the existing envelope item type, [`attachment`](https://develop.sentry.dev/sdk/envelopes/#attachment) to transport the video data.
```

Comment:
I believe we can't use the type 'attachment' because it will impact attachment quota

Reply:
Agree, I guess we need a new type something like `MobileReplayRecording`, similar to [ReplayRecording](https://github.com/getsentry/sentry/issues/63255)

Reply:
In that case I'd go with something less tied to being "mobile". For example even the Cocoa and Flutter SDKs could eventually work the same on desktop.

Instead, I'd suggest `ReplayVideoRecording`, assuming there's `ReplayRecording` (where does that come from? the linked page doesn't mention it)

Reply:
hm, yeah looks outdate 🤔 here they are https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-server/src/envelope.rs#L112-L115

`ReplayVideoRecording` sounds better indeed!

Reply:
Here's the js sdk envelope creation if you're interested https://github.com/getsentry/sentry-javascript/blob/develop/packages/replay/src/util/createReplayEnvelope.ts#L15

Reply:
> hm, yeah looks outdated 🤔 here they are [getsentry/relay@`5fd3969`/relay-server/src/envelope.rs#L112-L115](https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-server/src/envelope.rs#L112-L115)

I wonder if we actually need another envelope item type or we can use `ReplayRecording`. This indicates it's just treated as "data": 
```
    /// ReplayRecordings, large blobs sent by the replay sdk
    ReplayRecordings,
```
https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-kafka/src/config.rs#L49

Reply:
Similarly, the JS code you've linked seems to take into account that it can be binary.

Reply:
> Here's the js sdk envelope creation if you're interested https://github.com/getsentry/sentry-javascript/blob/develop/packages/replay/src/util/createReplayEnvelope.ts#L15

I believe we're good to go just by copying this.

One item is ReplayEvent, another item is the video binary itself, the type could be an extension of the existing one like `replay_recording.video`.

Reply:
The RRWeb JSON should also work fine, with the `Custom` event type, it can have any data JSON object content. On UI end, it just needs a custom handler code that seemed reasonable to do: https://github.com/rrweb-io/rrweb/blob/master/docs/recipes/custom-event.md

The missing bit (for me) is how how reference the video sent in another envelope item. Like do we give it some pre-authenticated link directly to GCS? Or is it going to be streamed through sentry servers?

Reply:
The UI/player side will likely need some additional metadata associated with the video (start/end timestamps) - where can we send this?

also cc @cmanallen for relay info

Reply:
> same envelope, different envelope item (type)

Oh, yes, make sense.

> The UI/player side will likely need some additional metadata associated with the video (start/end timestamps) - where can we send this?

This goes in the RRWeb json new type I believe. Something like:
```json
[
    {
       "type": 4,
       "data": {
          "href": "",
          "width": 390,
          "height": 720
       },
      "timestamp": 1707254438643
    },
    {
       "type": 8,
       "data": {
          "segment_id": 4,
          "duration": 5000
        }
    }
]
 ```

Reply:
That sounds good -- the backend will need to ensure we can reference the mp4 via `segment_id`.

I would suggest using `type = 5` so that we don't need to override rrweb's enum:

```
{
   "type": 5,
   "data": {
      "tag": "video",
      "payload": {
        "segment_id": 4,
        "duration": 5000
      }
    }
}
```

Reply:
what about the existing `segment_id` on the [ReplayEvent](https://getsentry.github.io/relay/relay_event_schema/protocol/struct.Replay.html#structfield.segment_id) type? Is this something different?

Reply:
FYI I've updated the markdown with the changes discussed in this thread.

Reply:
added it there